File: backend/src/validators/doubled_transaction.py
from datetime import timedelta
from ..core.state import Store


def validate(store: Store, transaction_data: dict) -> list[str]:
    account = store.get_account()
    if account is None:
        return []
    
    current_time = transaction_data.get("time")
    if current_time is None:
        return []
    
    current_merchant = transaction_data.get("merchant", "")
    current_amount = transaction_data.get("amount", 0)
    
    two_minutes_ago = current_time - timedelta(minutes=2)
    
    for t in store.get_transactions():
        if t.time >= two_minutes_ago:
            if t.merchant == current_merchant and t.amount == current_amount:
                return ["doubled-transaction"]
    return []

File: frontend/src/core/validators/doubledTransaction.js
export function validate(store, transactionData) {
  const account = store.getAccount();
  if (account === null) {
    return [];
  }

  const currentTime = transactionData.time;
  if (!currentTime) {
    return [];
  }

  const currentMerchant = transactionData.merchant || '';
  const currentAmount = transactionData.amount || 0;

  const twoMinutesAgo = new Date(currentTime.getTime() - 2 * 60 * 1000);

  for (const t of store.getTransactions()) {
    if (t.time >= twoMinutesAgo) {
      if (t.merchant === currentMerchant && t.amount === currentAmount) {
        return ['doubled-transaction'];
      }
    }
  }
  return [];
}